src/router.py

```diff
@@ -0,0 +1,194 @@
+            mask_int = int.from_bytes(entry.mask, RoutingTable.ENDIANITY)
+            entry_subnet_int = int.from_bytes(socket.inet_aton(entry.subnet), RoutingTable.ENDIANITY)
+            if dest_ip_int & mask_int == entry_subnet_int:
+                return entry.interface_ip
+        return None
+    
+
+class Router:
+    """
+    Performs the main logic of a router. Forwards IP 
+    packets and responds to ARP requests """
+
+    DEFAULT = '0.0.0.0'
+    MASK_24_SUBNET = socket.inet_aton('255.255.255.0')
+    DEFAULT_MASK = socket.inet_aton('0.0.0.0')
+    NET1_NAME = 'net1'
+    NET2_NAME = 'net2'
+    NET1_IP = '1.1.1.1'
+    NET2_IP = '2.2.2.1'
+    CLIENT2_IP = '2.2.2.2'
+    CLIENT1_IP = '1.1.1.2'
+    
+    ETH_TYPE_ARP = b'\x08\x06'
+    ETH_TYPE_IPV4 = b'\x08\x00'
+    BUF_SIZE = 2048
+    GGP = 0x0003
+
+    MAC_ROUTER_NET1 = '02:42:36:E0:40:43'
+    MAC_ROUTER_NET2 = '02:42:00:33:82:46'
+    CLIENT1_MAC = '02:42:01:01:01:02'
+    CLIENT2_MAC = '02:42:02:02:02:02'
+
+    INTERFACE_MAC2NAME = {MAC_ROUTER_NET1: NET1_NAME,
+                        MAC_ROUTER_NET2: NET2_NAME}
+
+    INTERFACE_NAME2IP = {NET1_NAME: NET1_IP,
+                        NET2_NAME: NET2_IP}
+
+    def __init__(self):
+        Interface = namedtuple('Interface', ['ip', 'mac', 'name'])
+        net1 = Interface(socket.inet_aton(Router.NET1_IP), Router.MAC_ROUTER_NET1, Router.NET1_NAME)
+        net2 = Interface(socket.inet_aton(Router.NET2_IP), Router.MAC_ROUTER_NET2, Router.NET2_NAME)
+        self.interfaces = { net1.name: net1,
+                            net2.name: net2}
+        self.arp_tbl = {}
+        RouteEntry = namedtuple('RouteEntry', ['subnet', 'mask', 'interface_ip'])
+        routes = [RouteEntry(Router.DEFAULT, Router.DEFAULT_MASK, None),
+                  RouteEntry('1.1.1.0', Router.MASK_24_SUBNET, Router.NET1_IP),
+                  RouteEntry('2.2.2.0', Router.MASK_24_SUBNET, Router.NET2_IP)]
+        self.routing_tbl = RoutingTable(routes)
+
+    def handle_ether(self, ether_struct, interface):
+        '''
+        Handles Ethernet packets by their ethernet type.
+        Triggers the relevant router's activity, including sending 
+        response packet (if neccessary).
+
+        :param ether_struct: the entire deconstructed packet as a Layers hirarchy.
+        '''
+        if ether_struct.get_ether_type() == ArpLayer.ETHR_TYPE:
+            response_packet = self.handle_arp(ether_struct, interface)
+            interface = interface  
+        elif ether_struct.get_ether_type() == IPv4Layer.ETHR_TYPE:
+            response_packet = self.handle_ip(ether_struct, interface)
+            if response_packet is None:
+                return
+            src_mac = MacAddress.mac2str(response_packet.get_src())
+            interface = self.INTERFACE_MAC2NAME[src_mac]
+        
+        if response_packet is None:
+            return
+        with socket.socket(socket.AF_PACKET, socket.SOCK_RAW, socket.ntohs(Router.GGP)) as raw_socket:
```

Comment:
Why create a new socket on every packet?
You could create a mapping between interfaces and pre-made sockets, avoiding the need for creating and binding a new socket every time.

This could save you time in handling packets